### PR TITLE
Reduced resolution loss

### PR DESCRIPTION
**Pull request by Andrewwango**
The reduced resolution loss for blur/downsampling problems which uses the measurement as the GT and a redowngraded image as the input, following e.g. Irani et al., Scanvic et al., and in pansharpening the classic reduced resolution "Wald" protocol.

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: deepinv/loss/__init__.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 from .mc import MCLoss
 from .ei import EILoss
 from .moi import MOILoss, MOEILoss
-from .sup import SupLoss
+from .sup import SupLoss, ReducedResolutionLoss
```
Comment (jscanvic) [inline deepinv/loss/__init__.py:4]: I'm happy with leaving it as is but it might be misleading to put it in `deepinv.loss.sup` when it's not really a supervised loss

File: deepinv/loss/sup.py [hunk 2]
```diff
@@ -41,3 +43,80 @@ def forward(self, x_net, x, **kwargs):
...
+        to the measurements :math:`y`.
+
+    .. info::
+
+        During training, consider using the `disable_train_metrics` option in :class:`deepinv.Trainer` to prevent a shape
+        mismatch during metric computation since the reduced resolution output will smaller than ground truth.
+
+    :param Metric, torch.nn.Module metric: metric used for computing data consistency,
+        which is set as the mean squared error by default.
+    """
```
Comment (jscanvic) [inline deepinv/loss/sup.py:80]: Add a reference to Irani's paper?

File: deepinv/loss/sup.py [hunk 3]
```diff
@@ -41,3 +43,80 @@ def forward(self, x_net, x, **kwargs):
...
+    class ReducedResolutionModel(Reconstructor):
+        def __init__(self, model):
+            super().__init__()
+            self.model = model
+
+        def forward(self, y, physics, **kwargs):
+            with torch.set_grad_enabled(self.training):
+                if self.training:
+                    try:
+                        y_rr = physics.A(y)
```
Comment (jscanvic) [inline deepinv/loss/sup.py]: In Michaeli's paper they don't have noise but in ours for deblurring we have noise and we use the noise model to generate the doubly corrupted measurements. Might make more sense to use `physics(y)` instead, or adding an option to compute either version. 

Also: `z` instead of `y_rr`?

File: deepinv/loss/sup.py [hunk 4]
```diff
@@ -41,3 +43,80 @@ def forward(self, x_net, x, **kwargs):
...
+    .. info::
+
+        During training, consider using the `disable_train_metrics` option in :class:`deepinv.Trainer` to prevent a shape
+        mismatch during metric computation since the reduced resolution output will smaller than ground truth.
+
+    :param Metric, torch.nn.Module metric: metric used for computing data consistency,
+        which is set as the mean squared error by default.
+    """
+
+    def forward(self, x_net, y, *args, **kwargs):
```
Comment (jscanvic) [inline deepinv/loss/sup.py]: Random thought: there's this pattern in the losses where we either expect the caller to feed in the pre-computed tensors (x_net) or we compute them in the loss directly (y_rr). Arguably, the 1st pattern enables more global optimizations (sharing variables between multiple loss for x_net, parallelization in data loading for y_rr) while the 2nd pattern is more convenient (we don't have to deal with computing stuff). I think it'd be nice to always have the choice to do one or the other. Not sure how though!

Comment (jscanvic) [inline deepinv/loss/sup.py]: Typing?

File: deepinv/loss/sup.py [hunk 5]
```diff
@@ -41,3 +43,80 @@ def forward(self, x_net, x, **kwargs):
...
+            return model
+        else:
+            return self.ReducedResolutionModel(model)
+
+    class ReducedResolutionModel(Reconstructor):
+        def __init__(self, model):
+            super().__init__()
+            self.model = model
+
+        def forward(self, y, physics, **kwargs):
```
Comment (jscanvic) [inline deepinv/loss/sup.py]: Typing?

File: deepinv/training/trainer.py [hunk 6]
```diff
@@ -691,6 +694,9 @@ def compute_metrics(
         :param int epoch: current epoch.
         :returns: The logs with the metrics.
         """
+        if train and self.disable_train_metrics:
```
Comment (jscanvic) [inline deepinv/training/trainer.py]: I'm happy with leaving it as is but I'd prefer guarding the calls instead of hijacking the behavior of the function

File: deepinv/training/trainer.py [hunk 7]
```diff
@@ -117,6 +117,8 @@ class Trainer:
     :param Metric, list[Metric] metrics: Metric or list of metrics used for evaluating the model.
         They should have ``reduction=None`` as we perform the averaging using :class:`deepinv.utils.AverageMeter` to deal with uneven batch sizes.
         :ref:`See the libraries' evaluation metrics <metric>`. Default is :class:`PSNR <deepinv.loss.metric.PSNR>`.
+    :param bool disable_train_metrics: if `True`, do not compute metrics on train set. Useful if your model output during training
+        is in the wrong shape for metric computation.
```
Comment (jscanvic) [inline deepinv/training/trainer.py:122]: Can you elaborate on what's going on here? I'm not entirely sure

File: deepinv/tests/test_loss.py [hunk 8]
```diff
@@ -666,3 +677,41 @@ def test_stacked_loss(device, imsize):
...
+        physics = dinv.physics.BlurFFT(
+            x.shape[1:], filter=dinv.physics.blur.gaussian_blur(0.4), device=device
+        )
+    else:
+        raise ValueError()
+
+    y = physics(x)
+
+    model.eval()
+    x_hat_eval = model(y, physics)  # just asjoint
```
Comment (jscanvic) [inline deepinv/tests/test_loss.py]: adjoint*?

File: deepinv/tests/test_loss.py [hunk 9]
```diff
@@ -666,3 +677,41 @@ def test_stacked_loss(device, imsize):
...
+    elif physics_name == "blur":
+        physics = dinv.physics.Blur(
+            filter=dinv.physics.blur.gaussian_blur(0.4), device=device
+        )
+    elif physics_name == "blurfft":
+        physics = dinv.physics.BlurFFT(
+            x.shape[1:], filter=dinv.physics.blur.gaussian_blur(0.4), device=device
+        )
+    else:
+        raise ValueError()
```
Comment (jscanvic) [inline deepinv/tests/test_loss.py]: Add a message?

File: deepinv/tests/test_loss.py [hunk 10]
```diff
@@ -666,3 +677,41 @@ def test_stacked_loss(device, imsize):
...
+
+    model.eval()
+    x_hat_eval = model(y, physics)  # just asjoint
+    assert x_hat_eval.shape == x.shape
+    assert metric(x_hat_eval, x) < 50
+
+    model.train()
+    x_hat_train = model(y, physics)
+    assert x_hat_train.shape == y.shape
+    assert metric(x_hat_train, y) < 50
```
Comment (jscanvic) [inline deepinv/tests/test_loss.py:723]: I'm not sure I understand the `metric(...) < 50`. What is it we're testing? That it doesn't work too well?

File: deepinv/tests/test_loss.py [hunk 11]
```diff
@@ -666,3 +677,41 @@ def test_stacked_loss(device, imsize):
...
+        physics = dinv.physics.Downsampling(filter=None, factor=2, device=device)
+    elif physics_name == "downsamplingmatlab":
+        physics = dinv.physics.DownsamplingMatlab(factor=2, device=device)
+    elif physics_name == "blur":
+        physics = dinv.physics.Blur(
+            filter=dinv.physics.blur.gaussian_blur(0.4), device=device
+        )
+    elif physics_name == "blurfft":
+        physics = dinv.physics.BlurFFT(
+            x.shape[1:], filter=dinv.physics.blur.gaussian_blur(0.4), device=device
```
Comment (jscanvic) [inline deepinv/tests/test_loss.py:706]: Just wondering, could we leave out the `device=device` in every condition and have `physics.to(device)` after the conditional?

File: deepinv/tests/test_loss.py [hunk 12]
```diff
@@ -319,15 +322,22 @@ def test_losses(
...
 
     save_dir = tmp_path / "dataset"
     # choose backbone denoiser
     backbone = dinv.models.AutoEncoder(
         dim_input=imsize[0] * imsize[1] * imsize[2], dim_mid=64, dim_hid=16
-    ).to(device)
+    )
+
+    if loss_name == "reducedresolution":
+        backbone = dinv.models.UNet(1, 1, scales=2)
```
Comment (jscanvic) [inline deepinv/tests/test_loss.py:341]: Can you elaborate on why it's needed?

File: deepinv/tests/test_loss.py [hunk 13]
```diff
@@ -319,15 +322,22 @@ def test_losses(
...
 
     # choose a reconstruction architecture
-    model = dinv.models.ArtifactRemoval(backbone, device=device)
+    model = dinv.models.ArtifactRemoval(backbone.to(device), device=device)
```
Comment (jscanvic) [inline deepinv/tests/test_loss.py]: Again, just wondering, could we leave out the `to(device)` and the `device=device` and add a `model.to(device)` below?

File: deepinv/loss/sup.py [hunk 14]
```diff
@@ -41,3 +43,80 @@ def forward(self, x_net, x, **kwargs):
...
+            with torch.set_grad_enabled(self.training):
```
Comment (jscanvic) [inline deepinv/loss/sup.py]: Does it make sense to disable grad here specifically?

File: deepinv/loss/sup.py [hunk 15]
```diff
@@ -41,3 +43,80 @@ def forward(self, x_net, x, **kwargs):
         :return: (:class:`torch.Tensor`) loss.
         """
         return self.metric(x_net, x)
+
+
+class ReducedResolutionLoss(SupLoss):
```
Comment (jscanvic) [inline deepinv/loss/sup.py:52]: Creative name - I like it!

File: deepinv/loss/sup.py [hunk 16]
```diff
@@ -41,3 +43,80 @@ def forward(self, x_net, x, **kwargs):
...
+    r"""
+    Reduced resolution loss for blur and downsampling problems.
+
+    The reduced resolution loss is defined as
+
+    .. math::
+
+        \frac{1}{n}\|y-\inverse{\forw{y}}\|^2
+
+    where :math:`\forw{y}` is the further-degraded measurement, and the measurement :math:`y` is used a supervisory signal.
```
Comment (jscanvic) [inline deepinv/loss/sup.py]: Happy to leave it as is but if we go down the "reduced resolution" route, we might want to emphasize resolution reduction in the docstrings more than the idea of twice measuring? I feel like "measurement" might be too broad here

**Review by jscanvic — COMMENTED** [review]
Thanks for the great feature! A few comments

File: deepinv/loss/__init__.py (hunk 1 above)
Comment (Andrewwango) [reply deepinv/loss/__init__.py:4]: true but mechanistically it's sup I guess

File: deepinv/loss/sup.py (hunk 4 above)
Comment (Andrewwango) [reply deepinv/loss/sup.py]: Hmm I agree and there were two design choices. IMO the biggest flaw in the second approach is to waste the original `x_net` which is always computed, and perform another forward pass. Am I right in thinking this?

File: deepinv/training/trainer.py (hunk 6 above)
Comment (Andrewwango) [reply deepinv/training/trainer.py]: What do you mean by guarding?

File: deepinv/tests/test_loss.py (hunk 10 above)
Comment (Andrewwango) [reply deepinv/tests/test_loss.py:723]: That the x_hat_train is a `f(A(y))` where `f = A^T` i.e. a degraded version of `y`, testing that the loss adapted model actually performs degradataion

File: deepinv/training/trainer.py (hunk 6 above)
Comment (jscanvic) [reply deepinv/training/trainer.py]: Instead of

```
def elsewhere(...):
    val = compute_metrics(...)

def compute_metrics(...):
    if ...:
       return dummy_value
    pass
```

have


```
def elsewhere(...):
    if ...:
       compute_metrics(...)

def compute_metrics(...):
    pass
```

Lemme know if that doesn't make sense lol

File: deepinv/loss/sup.py (hunk 4 above)
Comment (jscanvic) [reply deepinv/loss/sup.py]: Exactly

File: deepinv/tests/test_loss.py (hunk 11 above)
Comment (Andrewwango) [reply deepinv/tests/test_loss.py:706]: I guess but it's not a big deal I don't think

File: deepinv/loss/sup.py (hunk 2 above)
Comment (Andrewwango) [reply deepinv/loss/sup.py:80]: done

File: deepinv/loss/sup.py (hunk 3 above)
Comment (Andrewwango) [reply deepinv/loss/sup.py]: done

File: deepinv/loss/sup.py (hunk 4 above)
Comment (Andrewwango) [reply deepinv/loss/sup.py]: Not sure if this is needed

Comment (Andrewwango) [reply deepinv/loss/sup.py]: done

File: deepinv/loss/sup.py (hunk 5 above)
Comment (Andrewwango) [reply deepinv/loss/sup.py]: done

File: deepinv/training/trainer.py (hunk 6 above)
Comment (Andrewwango) [reply deepinv/training/trainer.py]: done, thanks!

File: deepinv/tests/test_loss.py (hunk 8 above)
Comment (Andrewwango) [reply deepinv/tests/test_loss.py]: done

File: deepinv/tests/test_loss.py (hunk 9 above)
Comment (Andrewwango) [reply deepinv/tests/test_loss.py]: done

File: deepinv/tests/test_loss.py (hunk 12 above)
Comment (Andrewwango) [reply deepinv/tests/test_loss.py:341]: done

File: deepinv/tests/test_loss.py (hunk 13 above)
Comment (Andrewwango) [reply deepinv/tests/test_loss.py]: done

File: deepinv/loss/sup.py (hunk 14 above)
Comment (Andrewwango) [reply deepinv/loss/sup.py]: done, gthanks

File: deepinv/loss/sup.py (hunk 16 above)
Comment (Andrewwango) [reply deepinv/loss/sup.py]: done

File: deepinv/training/trainer.py (hunk 7 above)
Comment (Andrewwango) [reply deepinv/training/trainer.py:122]: doen

File: deepinv/training/trainer.py [hunk 17]
```diff
@@ -808,8 +813,12 @@ def step(
...
             # Log metrics
-            logs = self.compute_metrics(
-                x, x_net, y, physics_cur, logs, train=train, epoch=epoch
+            logs = (
+                logs
+                if self.disable_train_metrics and train
+                else self.compute_metrics(
+                    x, x_net, y, physics_cur, logs, train=train, epoch=epoch
+                )
             )
```
Comment (jscanvic) [inline deepinv/training/trainer.py]: Sorry for the nitpick

```suggestion
            if not (self.disable_train_metrics and train):
                logs = self.compute_metrics(
                    x, x_net, y, physics_cur, logs, train=train, epoch=epoch
                )
```

File: deepinv/tests/test_loss.py (hunk 10 above)
Comment (jscanvic) [reply deepinv/tests/test_loss.py:723]: Oh so that's what DummyModel does, gotcha!

File: deepinv/tests/test_loss.py (hunk 11 above)
Comment (jscanvic) [reply deepinv/tests/test_loss.py:706]: Gotcha!
